contracts/DomainRegistry.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "hardhat/console.sol";
+import "./Ownable.sol";
+import "./DomainValidator.sol";
+
+contract DomainRegistry is Ownable {
+    uint256 private totalReservedDomains;
+    uint256 private minDeposit;
+
+    struct Domain {
+        address controller;
+        uint256 deposit;
+    }
+
+    mapping(string => Domain) private domains;
+
+    event DomainReserved(string indexed domainName, address indexed controller, uint256 deposit);
+    event DepositChanged(string indexed domainName, uint256 newDeposit);
+    event DomainControlTransferred(string indexed domainName, address indexed newController);
+    event DomainReleased(string indexed domainName, address indexed controller, uint256 refundAmount);
+
+    constructor(uint256 _minDeposit) payable  {
+        minDeposit = _minDeposit;
+    }
+
+    modifier onlyController(string memory domainName) {
+        require(domains[domainName].controller == msg.sender, "Only the domain controller can execute it");
+        _;
+    }
+
+    modifier sufficientDeposit(uint256 amount) {
+        require(msg.value >= amount, "Insufficient deposit amount");
+        _;
+    }
+
+    function reserveDomain(string memory domainName) public payable  {
+        string memory cleanDomainName = DomainValidator.stripProtocolOnlyAssembly(domainName);
+        require(DomainValidator.isValidDomain(cleanDomainName), "Invalid domain name format");
+        require(hasValidParentDomain(cleanDomainName), "Parent domain must exist");
+        require(domains[cleanDomainName].controller == address(0), "Domain already reserved");
+
+        domains[cleanDomainName] = Domain({
+            controller: msg.sender,
+            deposit: msg.value 
+        });
+        totalReservedDomains++;
+
+        emit DomainReserved(cleanDomainName, msg.sender, msg.value);
+    }
+
+    function reserveDomainOnlyAssymbly(string memory domainName) public payable  {
+        string memory cleanDomainName = DomainValidator.stripProtocolOnlyAssembly(domainName);
+        require(DomainValidator.isValidDomain(cleanDomainName), "Invalid domain name format");
+        require(hasValidParentDomain(cleanDomainName), "Parent domain must exist");
+        require(domains[cleanDomainName].controller == address(0), "Domain already reserved");
+
+        domains[cleanDomainName] = Domain({
+            controller: msg.sender,
+            deposit: msg.value 
+        });
+        totalReservedDomains++;
+
+        emit DomainReserved(cleanDomainName, msg.sender, msg.value);
+    }
```

Comment:
No reason to bring both these functions to the contract. Figure out a better way for gas efficiency testing. Take into account that only one version will go to production (with or without assemblies depending on technical decision)